notebooks/generative-ai/question-answering.ipynb

```diff
@@ -208,43 +244,375 @@
+        "\n",
+        "----\n",
+        "{context}\n",
+        "----\n",
+        "Question: {question}\n",
+        "\"\"\"\n",
+        ")\n",
+        "\n",
+        "DOCUMENT_PROMPT = PromptTemplate.from_template(\"\"\"\n",
+        "---\n",
+        "NAME: {name}\n",
+        "PASSAGE:\n",
+        "{page_content}\n",
+        "---\n",
+        "\"\"\")\n",
+        "\n",
+        "CONDENSE_QUESTION_PROMPT = PromptTemplate.from_template(\n",
+        "\"\"\"Given the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
+        "\n",
+        "Chat History:\n",
+        "{chat_history}\n",
+        "Follow Up Input: {question}\n",
+        "Standalone question:\n",
+        "\"\"\")\n",
+        "\n",
+        "standalone_question = RunnableMap(\n",
+        "    standalone_question=RunnablePassthrough.assign(\n",
+        "        chat_history=lambda x: _format_chat_history(x[\"chat_history\"])\n",
+        "    )\n",
+        "    | CONDENSE_QUESTION_PROMPT\n",
+        "    | llm\n",
+        "    | StrOutputParser(),\n",
+        ")\n",
+        "\n",
+        "def _format_chat_history(chat_history) -> str:\n",
+        "    buffer = \"\"\n",
+        "    for dialogue_turn in chat_history:\n",
+        "        human = \"Human: \" + dialogue_turn[0]\n",
+        "        ai = \"Assistant: \" + dialogue_turn[1]\n",
+        "        buffer += \"\\n\" + \"\\n\".join([human, ai])\n",
+        "    return buffer\n",
+        "\n",
+        "def _combine_documents(\n",
+        "    docs, document_prompt=DOCUMENT_PROMPT, document_separator=\"\\n\\n\"\n",
+        "):\n",
+        "    doc_strings = [format_document(doc, document_prompt) for doc in docs]\n",
+        "    return document_separator.join(doc_strings)\n",
+        "\n",
+        "retrieved_documents = RunnableMap(\n",
+        "    docs=itemgetter(\"standalone_question\") | retriever,\n",
+        "    question=itemgetter(\"standalone_question\"),\n",
+        ")\n",
+        "\n",
+        "_context = {\n",
+        "    \"context\": lambda x: _combine_documents(x[\"docs\"]),\n",
+        "    \"question\": lambda x: x[\"question\"],\n",
+        "}\n",
+        "\n",
+        "answer = {\n",
+        "    \"answer\": _context | ANSWER_PROMPT | llm,\n",
+        "    \"docs\": itemgetter(\"docs\"),\n",
+        "}\n",
+        "\n",
+        "chain = (\n",
+        "    standalone_question | retrieved_documents | answer\n",
+        ")\n",
+        "\n",
+        "ans = chain.invoke({ \n",
+        "  \"question\": \"What are their objectives?\", \n",
+        "  \"chat_history\": [\n",
+        "    \"What is the nasa sales team?\",\n",
+        "    \"The sales team of NASA consists of Laura Martinez, the Area \"\n",
+        "    \"Vice-President of North America, and Gary Johnson, the Area \"\n",
+        "    \"Vice-President of South America. (Sales Organization Overview)\",\n",
```

Comment:
Should this say `(SOURCE: Sales Organization Overview)` to be consistent with the responses? Or else remove the `SOURCE:` prefixes in the example output?

Note that I'm not running the code, just looking at the logic and the outputs that you saved.

Reply:
good spot, updated.